File: tasks/task4/booking-service/MonolithRestClients/PromoClient.cs
using System.Text.Json;

namespace Hotelio.BookingService.MonolithRestClients
{
    public interface IPromoClient
    {
        Task<PromoValidationResult?> ValidatePromoCode(string promoCode, string userId);
    }

    public class PromoValidationResult
    {
        public bool IsValid { get; set; }
        public double DiscountAmount { get; set; }
        public string? PromoCode { get; set; }
    }

    public class PromoClient : IPromoClient
    {
        private readonly HttpClient _httpClient;
        private readonly ILogger<PromoClient> _logger;

        public PromoClient(HttpClient httpClient, ILogger<PromoClient> logger)
        {
            _httpClient = httpClient;
            _logger = logger;
        }

        public async Task<PromoValidationResult?> ValidatePromoCode(string promoCode, string userId)
        {
            try
            {
                var url = $"/api/promos/validate?code={Uri.EscapeDataString(promoCode)}&userId={Uri.EscapeDataString(userId)}";
                var response = await _httpClient.PostAsync(url, null);
                response.EnsureSuccessStatusCode();

                var content = await response.Content.ReadAsStringAsync();
                var promo = JsonSerializer.Deserialize<PromoDto>(content, new JsonSerializerOptions
                {
                    PropertyNameCaseInsensitive = true
                });

                if (promo == null) return null;

                return new PromoValidationResult
                {
                    IsValid = true,
                    DiscountAmount = promo.Discount,
                    PromoCode = promo.Code
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to validate promo code {PromoCode} for user {UserId}",
                    promoCode, userId);
                return null;
            }
        }

        private class PromoDto
        {
            public string? Code { get; set; }
            public double Discount { get; set; }
        }
    }
}

File: tasks/task5/booking-service/rest-clients/promoClient.js
const axios = require('axios');

const PROMO_SERVICE_URL = process.env.PROMO_SERVICE_URL || 'http://monolith:8080';

class PromoClient {
    async validatePromo(promoCode, userId) {
        if (!promoCode) return null;

        try {
            const response = await axios.post(
                `${PROMO_SERVICE_URL}/api/promos/validate?code=${promoCode}&userId=${userId}`
            );
            return {
                code: response.data.code,
                discountPercent: response.data.discountPercent,
                active: response.data.active,
                vipOnly: response.data.vipOnly
            };
        } catch (error) {
            if (error.response?.status === 400) {
                return null;
            }
            throw new Error(`Failed to validate promo: ${error.message}`);
        }
    }
}

module.exports = new PromoClient();